osquery/tables/system/windows/logon_sessions.cpp

```diff
@@ -0,0 +1,195 @@
+
+LONGLONG filetimeToUnixtime(const LARGE_INTEGER& ft) {
+  LARGE_INTEGER date, adjust;
+  date.HighPart = ft.HighPart;
+  date.LowPart = ft.LowPart;
+  adjust.QuadPart = 11644473600000 * 10000;
+  date.QuadPart -= adjust.QuadPart;
+  return date.QuadPart / 10000000;
+}
+
+LPTSTR GetLogonType(ULONG logonType) {
+  switch (logonType) {
+  case 2:
+    return TEXT("Interactive");
+  case 3:
+    return TEXT("Network");
+  case 4:
+    return TEXT("Batch");
+  case 5:
+    return TEXT("Service");
+  case 6:
+    return TEXT("Proxy");
+  case 7:
+    return TEXT("Unlock");
+  case 8:
+    return TEXT("NetworkCleartext");
+  case 9:
+    return TEXT("NewCredentials");
+  case 10:
+    return TEXT("RemoteInteractive");
+  case 11:
+    return TEXT("CachedInteractive");
+  case 12:
+    return TEXT("CachedRemoteInteractive");
+  case 13:
+    return TEXT("CachedUnlock");
+  default:
+    return TEXT("None");
+  }
+}
+
+QueryData QueryLogonSessions(QueryContext& context) {
+  QueryData results;
+  bool loadedManually = false;
+
+  HMODULE module = GetModuleHandle(TEXT("Secur32.dll"));
+
+  if (!module) {
+    module = LoadLibrary(TEXT("Secur32.dll"));
+    loadedManually = true;
+  }
+
+  NTSTATUS(__stdcall * LsaEnumerateLogonSessions)
+  (PULONG LogonSessionCount, PLUID * LogonSessionList);
+  LsaEnumerateLogonSessions =
+      reinterpret_cast<decltype(LsaEnumerateLogonSessions)>(
+          GetProcAddress(module, "LsaEnumerateLogonSessions"));
+
+  NTSTATUS(__stdcall * LsaGetLogonSessionData)
+  (PLUID LogonId, PSECURITY_LOGON_SESSION_DATA * ppLogonSessionData);
+  LsaGetLogonSessionData = reinterpret_cast<decltype(LsaGetLogonSessionData)>(
+      GetProcAddress(module, "LsaGetLogonSessionData"));
+
+  NTSTATUS(__stdcall * LsaFreeReturnBuffer)(PVOID Buffer);
+  LsaFreeReturnBuffer = reinterpret_cast<decltype(LsaFreeReturnBuffer)>(
+      GetProcAddress(module, "LsaFreeReturnBuffer"));
+
+  if (LsaEnumerateLogonSessions && LsaGetLogonSessionData &&
+      LsaFreeReturnBuffer) {
+    PLUID sessions = NULL;
+    ULONG sessionCount = 0;
+    NTSTATUS status = LsaEnumerateLogonSessions(&sessionCount, &sessions);
+
+    for (size_t i = 0; i < sessionCount; i++) {
+      PSECURITY_LOGON_SESSION_DATA sessionData = NULL;
+      NTSTATUS status = LsaGetLogonSessionData(&sessions[i], &sessionData);
+
+      // I need to go back and convert LARGE_INTEGER values to Unix Timestamp
+      if (status == 0) {
+        Row r;
+        r["logon_id"] = INTEGER(sessionData->LogonId.LowPart);
+        r["user"] = wstringToString(sessionData->UserName.Buffer);
+        r["logon_domain"] = wstringToString(sessionData->LogonDomain.Buffer);
+        r["authentication_package"] =
+            wstringToString(sessionData->AuthenticationPackage.Buffer);
+        r["logon_type"] = GetLogonType(sessionData->LogonType);
+        r["session_id"] = INTEGER(sessionData->Session);
+        LPTSTR sid;
+        ConvertSidToStringSid(sessionData->Sid, &sid);
```

Comment:
This has a return value of BOOL, you should check it. If it fails, the next line will be accessing an invalid pointer.